File: haskell/src/Solutions/Euler136.hs
module Solutions.Euler136 where
import Utils.NumberTheory(unPrime, primesUpTo)

bound = 49999999 :: Int

take4s p = takeWhile (\x -> p * (4^x) <= bound) [if p `mod` 4 == 3 then 0 else 1..2]

euler136 :: IO String
euler136 = return $ show $ (+2) $ sum $ map (length . take4s . unPrime) $ tail (primesUpTo bound)

File: rust/src/solutions/euler136.rs
use std::ops::Add;

pub fn euler136() -> String {
    let bound = 50_000_000;
    primal::Sieve::new(bound)
        .primes_from(3)
        .take_while(|p| p < &bound)
        .map(|p| {
            (if p % 4 == 3 { 0 } else { 1 }..3)
                .take_while(|&x| p * 4usize.pow(x) <= bound)
                .count()
        })
        .sum::<usize>()
        .add(2)
        .to_string()
}
